dex/networks/zec/tx.go

```diff
@@ -0,0 +1,417 @@
+		return nil, fmt.Errorf("only version 4 (sapling) supported")
+	}
+
+	err := putUint32(w, uint32(tx.Version)|(1<<31))
+	if err != nil {
+		return nil, err
+	}
+
+	err = putUint32(w, versionSaplingGroupID)
+	if err != nil {
+		return nil, err
+	}
+
+	count := uint64(len(tx.MsgTx.TxIn))
+	err = wire.WriteVarInt(w, pver, count)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, ti := range tx.TxIn {
+		err = writeTxIn(w, tx.Version, ti)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	count = uint64(len(tx.TxOut))
+	err = wire.WriteVarInt(w, pver, count)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, to := range tx.TxOut {
+		err = wire.WriteTxOut(w, pver, tx.Version, to)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err = putUint32(w, tx.LockTime); err != nil {
+		return nil, err
+	}
+
+	if err = putUint32(w, tx.ExpiryHeight); err != nil {
+		return nil, err
+	}
+
+	if tx.Version == VersionSapling {
+		// valueBalance
+		if err := putUint64(w, 0); err != nil {
+			return nil, err
+		}
+
+		// nShieldedSpend
+		err = wire.WriteVarInt(w, pver, 0)
+		if err != nil {
+			return nil, err
+		}
+
+		// nShieldedOutput
+		err = wire.WriteVarInt(w, pver, 0)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if err := wire.WriteVarInt(w, pver, 0); err != nil {
+		return nil, err
+	}
+
+	return w.Bytes(), nil
+}
+
+// see https://zips.z.cash/protocol/protocol.pdf section 7.1
+func DeserializeZecTx(b []byte) (*ZecTx, error) {
+	tx := &ZecTx{MsgTx: new(wire.MsgTx)}
+	r := bytes.NewReader(b)
+	if err := tx.ZecDecode(r); err != nil {
+		return nil, err
+	}
+	return tx, nil
+}
+
+// ZecDecode reads the serialized transaction from the reader and populates the
+// *ZecTx's fields.
+func (tx *ZecTx) ZecDecode(r io.Reader) error {
+	ver, err := readUint32(r)
+	if err != nil {
+		return err
+	}
+
+	// overWintered := (ver & (1 << 31)) > 0
+	ver &= overwinterMask // Clear the overwinter bit
+	tx.Version = int32(ver)
+
+	if ver > 4 {
+		return fmt.Errorf("unsupported tx version %d > 4", ver)
+	}
+
+	if ver > 3 {
+		_, err := readUint32(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	txInCount, err := wire.ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+
+	tx.TxIn = make([]*wire.TxIn, 0, txInCount)
+	for i := 0; i < int(txInCount); i++ {
+		ti := new(wire.TxIn)
+		if err := readTxIn(r, pver, int32(ver), ti); err != nil {
+			return err
+		}
+		tx.TxIn = append(tx.TxIn, ti)
+	}
+
+	txOutCount, err := wire.ReadVarInt(r, pver)
+	if err != nil {
+		return err
+	}
+
+	tx.TxOut = make([]*wire.TxOut, 0, txOutCount)
+	for i := 0; i < int(txOutCount); i++ {
+		to := new(wire.TxOut)
+		if err := readTxOut(r, pver, int32(ver), to); err != nil {
+			return err
+		}
+		tx.TxOut = append(tx.TxOut, to)
+	}
+
+	tx.LockTime, err = readUint32(r)
+	if err != nil {
+		return err
+	}
+
+	if ver > 3 {
+		tx.ExpiryHeight, err = readUint32(r)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Empty the rest of the buffer.
+	if ver < 2 {
+		return nil
```

Comment:
`// Empty the rest of the buffer.` meaning the rest of the buffer should be empty?  Or we should read and discard the rest before returning?  Or just, "done".

Reply:
I guess I just meant to discard the rest before returning. I've removed the confusing message now.